aws/s3_test.go

```diff
@@ -132,332 +163,277 @@ func (b TestS3Bucket) addObject(s SetupInfo, fileName string, fileBody string) e
-	bucketNamesPerRegion, err := getAllS3Buckets(awsSession, time.Now().Add(1*time.Hour*-1), targetRegions, s.bucketName, batchSize)
-	if batchSize < 0 {
-		if err == nil {
-			assert.Fail(t, "Did not fail for invalid batch size")
+		if err != nil {
+			assert.Failf(t, "Failed to list S3 Buckets", errors.WithStackTrace(err).Error())
 		}
-		return
-	}
 
-	if err != nil {
-		assert.Failf(t, "Failed to list S3 Buckets", errors.WithStackTrace(err).Error())
-	} else {
-		assert.NotContains(t, bucketNamesPerRegion[s.region], s.bucketName)
-	}
+		// Validate test bucket does not exist before creation
+		assert.NotContains(t, bucketNamesPerRegion[awsParams.region], aws.String(bucket.name))
 
-	// Create test bucket
-	bucket := TestS3Bucket{
-		name: s.bucketName,
-	}
-	if bucketTags != nil && len(bucketTags) > 0 {
-		bucket.tags = bucketTags
-	}
-	err = bucket.create(s.svc)
-	if err != nil {
-		assert.Failf(t, "Failed to create test bucket", errors.WithStackTrace(err).Error())
-	}
+		// Create test bucket
+		if args.bucketTags != nil && len(args.bucketTags) > 0 {
+			bucket.tags = args.bucketTags
+		}
+		err = bucket.create(awsParams.svc)
+		if err != nil {
+			assert.Failf(t, "Failed to create test bucket", errors.WithStackTrace(err).Error())
+		}
 
-	bucketNamesPerRegion, err = getAllS3Buckets(awsSession, time.Now().Add(1*time.Hour), targetRegions, s.bucketName, batchSize)
-	if err != nil {
-		assert.Failf(t, "Failed to list S3 Buckets", errors.WithStackTrace(err).Error())
-	}
+		bucketNamesPerRegion, err = getAllS3Buckets(awsSession, time.Now().Add(1*time.Hour), targetRegions, bucket.name, args.batchSize)
+		if err != nil {
+			assert.Failf(t, "Failed to list S3 Buckets", errors.WithStackTrace(err).Error())
+		}
 
-	if bucketTags != nil && len(bucketTags) > 0 {
-		for _, tag := range bucketTags {
-			key := strings.ToLower(tag["Key"])
-			value := strings.ToLower(tag["Value"])
-			if key == "cloud-nuke-excluded" && value == "true" {
-				assert.NotContains(t, bucketNamesPerRegion[s.region], aws.String(s.bucketName))
-			}
+		if args.shouldMatch {
+			assert.Contains(t, bucketNamesPerRegion[awsParams.region], aws.String(bucket.name))
+			t.Logf("SUCCESS: Matched bucket - %s", bucket.name)
+		} else {
+			assert.NotContains(t, bucketNamesPerRegion[awsParams.region], aws.String(bucket.name))
+			t.Logf("SUCCESS: Did not match bucket - %s", bucket.name)
 		}
-	}
+	}, nil
+}
 
-	if shouldMatch {
-		assert.Contains(t, bucketNamesPerRegion[s.region], aws.String(s.bucketName))
+// TestListS3Bucket tests listing S3 bucket operation
+func TestListS3Bucket(t *testing.T) {
+	type testCaseStruct struct {
+		name string
+		args TestListS3BucketArgs
 	}
-}
 
-func TestList_EmptyS3Bucket_NoTags(t *testing.T) {
-	testListS3BucketsWrapper(t, []map[string]string{}, 10, true)
-}
+	var allTestCases []testCaseStruct
 
-func TestList_EmptyS3Bucket_WithoutFilterTag(t *testing.T) {
-	testListS3BucketsWrapper(t, []map[string]string{
+	testCases := []testCaseStruct{
```

Comment:
NIT: The Go idiom typically used for table-driven tests declares the struct and the tests all in one:

```go
var testCases = []struct {
	param1 string
	param2 string
}{
	{"aaa", "bbb"},
	{"ccc", "ddd"},
}
```

Reply:
Done for `TestListS3Bucket` . For `TestNukeS3Bucket` as per - https://github.com/gruntwork-io/cloud-nuke/blob/7a99393a574793e7a55e16876bf0923ddc6c9845/aws/s3_test.go#L428 - as we need to generate tests both for versioned and non versioned buckets - we have to create list of tests first and then execute them.